src/metadata/firmware_download.rs

```diff
@@ -0,0 +1,150 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+// SPDX-FileCopyrightText: 2025 1BitSquared <info@1bitsquared.com>
+// SPDX-FileContributor: Written by Rachel Mant <git@dragonmux.network>
+// SPDX-FileContributor: Modified by P-Storm <pauldeman@gmail.com>
+
+use std::path::PathBuf;
+
+use color_eyre::eyre::ContextCompat;
+use serde::Deserialize;
+use url::Url;
+
+#[derive(Deserialize, Clone)]
+#[serde(deny_unknown_fields)]
+pub struct FirmwareDownload
+{
+	#[serde(rename = "friendlyName")]
+	pub friendly_name: String,
+	#[serde(rename = "fileName")]
+	pub file_name: PathBuf,
+	pub uri: Url,
+}
+
+impl FirmwareDownload
+{
+	pub fn calculate_documentation_url(&self) -> color_eyre::Result<Url>
+	{
+		let variant_uri = &self.uri;
+
+		// Convert the path compoment of the download URI to a Path
+		let mut docs_path = PathBuf::from(variant_uri.path());
+		// Replace the file extension from ".elf" to ".md"
+		docs_path.set_extension("md");
+		// Copy only the origin
+		let mut docs_uri = Url::parse(&variant_uri.origin().ascii_serialization())?;
+		docs_uri.set_path(docs_path.to_str().expect("Can't set a path from a doc path"));
+
+		Ok(docs_uri)
+	}
+
+	pub fn calculate_release_uri(&self, release: &str) -> color_eyre::Result<Url>
+	{
+		// Find where the release tag component is in the path, stripping back to that
+		let mut path_segments = self.uri.path_segments().context("cannot be base")?.collect::<Vec<_>>();
+
+		// Find the release segment position
+		let release_segment_position = path_segments
+			.iter()
+			.position(|s| s.ends_with(release))
+			.with_context(|| {
+				format!("This firmware URL doesn't contain the segment release with value '{}'", release)
+			})?;
+
+		let new_segments = path_segments
+			.as_mut_slice()
+			.get_mut(..=release_segment_position)
```

Comment:
Some explanation as to what the `..=` syntax even does would be appreciated - preferably as a comment and to indicate your intent here. What is this trying to do?

Reply:
The idea there is to get the allocated range up until the release segment position.  the mut part is because the `download` segment is getting changed into a `tag` segment.